REGRESSION: Scrolling down and attempting to fullscreen video on twitter.com displays the feed

pxlcoder opened the pull request:
#### e680442311f32c01dbf7187d032fcde640e1a525
<pre>
REGRESSION: Scrolling down and attempting to fullscreen video on twitter.com displays the feed
<a href="https://bugs.webkit.org/show_bug.cgi?id=269795">https://bugs.webkit.org/show_bug.cgi?id=269795</a>
<a href="https://rdar.apple.com/122981183">rdar://122981183</a>

Reviewed by Jer Noble.

In order for element fullscreen to behave correctly, sites rely on the following
invariants:

1. &quot;fullscreenchange&quot; is fired before &quot;resize&quot;.
2. The values of viewport properties during &quot;fullscreenchange&quot; match the fullscreen size.

Twitter relies on (1), as they teardown the video player if &quot;resize&quot; occurs
during the transition into fullscreen.

270199@main broke two things:

1. The ordering of &quot;fullscreenchange&quot; and &quot;resize&quot; events when entering fullscreen on iPadOS.
2. The values of viewport properties during the &quot;fullscreenchange&quot; event on visionOS.

(1) went undetected for a long time, because at first, the effect was not this
bug, but a crash, fixed in 273885@main.

272752@main fixed (2), but also introduced (1) on visionOS.

To the invariant mentioned above is true, fix by adding logic which ensures
&quot;resize&quot; is dispatched after &quot;fullscreenchange&quot;.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setNeedsDOMWindowResizeEvent):
(WebCore::Document::setNeedsVisualViewportResize):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::setAnimatingFullscreen):
(WebCore::FullscreenManager::addPendingScheduledResize):
* Source/WebCore/dom/FullscreenManager.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/FullscreenLayoutParameters.mm: Renamed from Tools/TestWebKitAPI/Tests/ios/FullscreenOverriddenLayoutParameters.mm.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/275115@main">https://commits.webkit.org/275115@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e75af5353e689cae478e4395539dfad9223c649d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40834 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19847 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43392 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36925 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43141 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22843 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17178 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33851 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41408 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16793 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35195 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14468 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14565 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36184 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44701 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37047 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36496 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40259 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15649 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12845 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38620 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17268 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9189 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17319 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16912 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5008,12 +5008,26 @@ void Document::updateViewportUnitsOnResize()
 void Document::setNeedsDOMWindowResizeEvent()
 {
+    if (CheckedPtr fullscreenManager = topDocument().fullscreenManagerIfExists()) {
+        if (fullscreenManager->isAnimatingFullscreen()) {
```
Style nit: prefer early returns to nested if.

pxlcoder replied:
I've removed the nesting.

jernoble commented on Source/WebCore/dom/FullscreenManager.h:
```diff
@@ -122,6 +128,8 @@ class FullscreenManager final : public CanMakeWeakPtr<FullscreenManager>, public
     Deque<GCReachableRef<Node>> m_fullscreenChangeEventTargetQueue;
     Deque<GCReachableRef<Node>> m_fullscreenErrorEventTargetQueue;
 
+    WeakHashMap<Document, OptionSet<ResizeType>, WeakPtrImplWithEventTargetData> m_documentsPendingScheduledResize;
```
A Fullscreen manager has a 1:1 relationship with the Document that owns it. So a WeakHashMap would be overkill here; it would only ever have a single Document in it. 

You should be able to simply make an OptionSet ivar here and use it in setAnimatingFullscreen() to determine if either event should be fired.

pxlcoder replied:
Done.

jernoble reviewed (approved):
R=me with one style nit and one functional nit.

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5008,12 +5008,26 @@ void Document::updateViewportUnitsOnResize()
 
 void Document::setNeedsDOMWindowResizeEvent()
 {
+    if (CheckedPtr fullscreenManager = topDocument().fullscreenManagerIfExists()) {
```
Needs to be wrapped in `#if ENABLE(FULLSCREEN_API)`.

pxlcoder committed:
REGRESSION: Scrolling down and attempting to fullscreen video on twitter.com displays the feed
https://bugs.webkit.org/show_bug.cgi?id=269795
rdar://122981183

Reviewed by Jer Noble.

In order for element fullscreen to behave correctly, sites rely on the following
invariants:

1. "fullscreenchange" is fired before "resize".
2. The values of viewport properties during "fullscreenchange" match the fullscreen size.

Twitter relies on (1), as they teardown the video player if "resize" occurs
during the transition into fullscreen.

270199@main broke two things:

1. The ordering of "fullscreenchange" and "resize" events when entering fullscreen on iPadOS.
2. The values of viewport properties during the "fullscreenchange" event on visionOS.

(1) went undetected for a long time, because at first, the effect was not this
bug, but a crash, fixed in 273885@main.

272752@main fixed (2), but also introduced (1) on visionOS.

To the invariant mentioned above is true, fix by adding logic which ensures
"resize" is dispatched after "fullscreenchange".

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setNeedsDOMWindowResizeEvent):
(WebCore::Document::setNeedsVisualViewportResize):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::setAnimatingFullscreen):
(WebCore::FullscreenManager::addPendingScheduledResize):
* Source/WebCore/dom/FullscreenManager.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/FullscreenLayoutParameters.mm: Renamed from Tools/TestWebKitAPI/Tests/ios/FullscreenOverriddenLayoutParameters.mm.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/275115@main
